spec/helpers/faker/puppet_module_remote_repo.rb

```diff
@@ -0,0 +1,130 @@
+require 'open3'
+
+require_relative '../faker'
+
+# Fake a remote git repository that hold a puppet module
+#
+# This module allow to fake a remote repositiory using:
+#  - a bare repo
+#  - a temporary cloned repo to operate on the remote before exposition to modulesync
+#
+# Note: This module need to have working_directory sets before using it
+module Faker
+  class PuppetModuleRemoteRepo
+    attr_reader :name
+    attr_reader :namespace
+
+    def initialize(name, namespace)
+      @name = name
+      @namespace = namespace
+    end
+
+    def populate
+      FileUtils.chdir(Faker.working_directory) do
+        run "git init --bare '#{bare_repo_dir}'"
+        run "git clone '#{bare_repo_dir}' '#{tmp_repo_dir}'"
+
+        module_short_name = name.sub(/^puppet-/, '')
+
+        FileUtils.chdir(tmp_repo_dir) do
+          metadata = {
+            name: "modulesync-#{module_short_name}",
+            version: '0.4.2',
+            author: 'ModuleSync team',
+          }.to_json
+
+          File.write 'metadata.json', metadata
+          run 'git add metadata.json'
+          run "git commit -m'Initial commit'"
+          run 'git push'
+        end
+      end
+    end
+
+    def read_only=(value)
+      mode = value ? '0444' : '0644'
+      FileUtils.chdir(bare_repo_dir) do
+        run "git config core.sharedRepository #{mode}"
+      end
+    end
+
+    def default_branch=(value)
+      FileUtils.chdir(bare_repo_dir) do
+        run "git branch -M master #{value}"
+        run "git symbolic-ref HEAD refs/heads/#{value}"
+      end
+    end
+
+    def read_file(filename, branch = nil)
+      FileUtils.chdir(tmp_repo_dir) do
+        run "git fetch"
+        run "git checkout #{branch}" unless branch.nil?
+        run "git merge --ff-only"
+        return unless File.exists?(filename)
+
+        return File.read filename
+      end
+    end
```

Comment:
This is odd to me. Why do you need to check out a branch to read a file? With pure git you can use `git show $ref:$filename`. It also returns an error code (128) if the file doesn't exist.

Reply:
Thanks, I did not catch this command! Plus, I can operate on the bare repo instead of the tmp, its awesome.

Reply:
Changes applied. Thanks.